cardano-api/src/Cardano/Api/Typed.hs

```diff
@@ -402,37 +429,129 @@ data StakeCredential
        = StakeCredentialByKey    (Hash StakeKey)
        | StakeCredentialByScript (Hash Script)
 
-data StakeCredentialReference
-       = StakeCredentialByValue   StakeCredential
-       | StakeCredentialByPointer StakeCredentialPointer
-       | NoStakeCredential
+data StakeAddressReference
+       = StakeAddressByValue   StakeCredential
+       | StakeAddressByPointer StakeAddressPointer
+       | NoStakeAddress
+
+type StakeAddressPointer = Shelley.Ptr
+
+
+instance HasTypeProxy (Address Byron) where
+    data AsType (Address Byron) = AsByronAddress
+    proxyToAsType _ = AsByronAddress
+
+
+instance HasTypeProxy (Address Shelley) where
+    data AsType (Address Shelley) = AsShelleyAddress
+    proxyToAsType _ = AsShelleyAddress
+
+
+instance HasTypeProxy StakeAddress where
+    data AsType StakeAddress = AsStakeAddress
+    proxyToAsType _ = AsStakeAddress
+
+
+instance SerialiseAsRawBytes (Address Byron) where
+    serialiseToRawBytes (ByronAddress addr) = CBOR.serialize' addr
+
+    deserialiseFromRawBytes AsByronAddress bs =
+      case CBOR.decodeFull' bs of
+        Left  _    -> Nothing
+        Right addr -> Just (ByronAddress addr)
+
+
+instance SerialiseAsRawBytes (Address Shelley) where
+    serialiseToRawBytes (ByronAddress addr) =
+        Shelley.serialiseAddr (Shelley.AddrBootstrap addr)
 
-type StakeCredentialPointer = Shelley.Ptr
+    serialiseToRawBytes (ShelleyAddress nw pc scr) =
+        Shelley.serialiseAddr (Shelley.Addr nw pc scr)
+
+    deserialiseFromRawBytes AsShelleyAddress bs =
+        case Shelley.deserialiseAddr bs of
+          Nothing -> Nothing
```

Comment:
Why not expose the error as an `Either` rather than a `Maybe`?

Reply:
:100: